### ISSUE TITLE
Reload page on updated hook (new files cached).

### ISSUE THREAD

**Issue by frederikhors** [yyx990803/register-service-worker#14]
I'm using the default file in a `cli-vue 3` new app.

What I need is to reload a page when service worker detects and updates new files.

Using your default hook I tried to use this code:

```js
updated (registration) {
    console.log('New content is available; please refresh.')
    window.location.reload()
  },
```

It works but it reloads forever!

**Why?**

I read a lot about Google Workbox and `@vue/cli-plugin-pwa` and I know these default hooks are bundled in final .js and executed on the main thread. So I think maybe I'm wrong?

I used `create-react-app` default `service-worker.js` and the reload used to work.

**Where am I wrong?**

**Comment by pranavjindal999** [comment]
I was experiencing the same issue and that was because for workbox, default value for `skipWaiting` is `false` now.

I added following to my `vue.config.js`

```js
 pwa: {
    workboxOptions: {
      skipWaiting: true
    }
  }
```

But that did not fix the issue either. Now the reload doesn't work at all. In fact none of the default hooks are called.

I even tried adding `clientsClaim: true` to `workboxOptions` but lo luck.

@yyx990803 if you could help ... 😄

**Comment by frederikhors** [comment]
@prana is there a way to debug in production all this?

**Comment by frederikhors** [comment]
@prana this can help? https://stackoverflow.com/questions/51214220/vue-cli-3-how-to-use-the-official-pwa-plugin-service-worker

**Comment by RedShift1** [comment]
You're supposed to call postMessage({action: 'skipWaiting'}) on the service worker but I can't figure out which variable it is bound to. Reloading won't help.

**Comment by pranavjindal999** [comment]
> I was experiencing the same issue and that was because for workbox, default value for `skipWaiting` is `false` now.
> 
> I added following to my `vue.config.js`
> 
> ```js
>  pwa: {
>     workboxOptions: {
>       skipWaiting: true
>     }
>   }
> ```
> 
> But that did not fix the issue either. Now the reload doesn't work at all. In fact none of the default hooks are called.
> 
> I even tried adding `clientsClaim: true` to `workboxOptions` but no luck.
> 
> @yyx990803 if you could help ... 😄

well somehow it worked for me automatically. I didn't change a thing.

**Comment by RedShift1** [comment]
> > I was experiencing the same issue and that was because for workbox, default value for `skipWaiting` is `false` now.
> > I added following to my `vue.config.js`
> > ```js
> >  pwa: {
> >     workboxOptions: {
> >       skipWaiting: true
> >     }
> >   }
> > ```
> > 
> > 
> > But that did not fix the issue either. Now the reload doesn't work at all. In fact none of the default hooks are called.
> > I even tried adding `clientsClaim: true` to `workboxOptions` but no luck.
> > @yyx990803 if you could help ... 😄
> 
> well somehow it worked for me automatically. I didn't change a thing.

I put this in vue.config.js (didn't have this file before):

```
module.exports =
    {
        pwa:
        {
            workboxOptions:
            {
                skipWaiting: true
            }
        }
    };
```
and automagically this seems to be working. Why this isn't documented better, who knows...

**Comment by ulivz** [comment]
See: [《How to Fix the Refresh Button When Using Service Workers》> Approach #3](https://redfin.engineering/how-to-fix-the-refresh-button-when-using-service-workers-a8e27af6df68)

**Comment by UnKnoWn-Consortium** [comment]
Workbox v4.1.0 has added a default message handler for `generateSW` output option that allows us to invoke the `skipWaiting` service worker method by `postMessage` from the window. It is available only when `skipWaiting` is false (i.e. the default value). 

We can leverage this with the version 4 of @vue/cli-plugin-pwa (which is, as of writing, still in its alpha). You can give it a try by `npm i @vue/cli-plugin-pwa@next`.

**Comment by OriginalEXE** [comment]
Indeed, as mentioned above, the alpha version of `@vue/cli-plugin-pwa` uses an updated workbox library which provides by default a listener that triggers the `skipWaiting` inside the service worker.

**Comment by shivanaru** [comment]
> > I was experiencing the same issue and that was because for workbox, default value for `skipWaiting` is `false` now.
> > I added following to my `vue.config.js`
> > ```js
> >  pwa: {
> >     workboxOptions: {
> >       skipWaiting: true
> >     }
> >   }
> > ```
> > 
> > 
> > But that did not fix the issue either. Now the reload doesn't work at all. In fact none of the default hooks are called.
> > I even tried adding `clientsClaim: true` to `workboxOptions` but no luck.
> > @yyx990803 if you could help ... 😄
> 
> well somehow it worked for me automatically. I didn't change a thing.

@pranavjindal999 - did you have to do: postMessage({action: 'skipWaiting'}) also? Or just the skipWaiting:true in the vue.config.js? Thanks!

**Comment by daneren2005** [comment]
Reading the above I wasn't sure how exactly to trigger the skip waiting event.  The syntax required to trigger the skipWaiting event with the default service worker is: 
```
registration.waiting.postMessage({
	type: 'SKIP_WAITING'
});
```

I also use this to hook into when the update is ready for the user's browser to refresh:
```
navigator.serviceWorker.addEventListener('controllerchange', () => {
	if(this.refreshing) {
		return;
	}
			
	this.refreshing = true;
	window.location.reload();
});
```

**Comment by Uysim** [comment]
Anyone have solution for this yet?

**Comment by IlyaSemenov** [comment]
@Uysim for @vue/cli-plugin-pwa this works in `vue.config.js`

```js
module.exports = {
	pwa: {
		workboxOptions: {
			// https://github.com/yyx990803/register-service-worker/issues/14
			// https://stackoverflow.com/questions/54145735/vue-pwa-not-getting-new-content-after-refresh
			skipWaiting: true,
		},
	},
}
```

for both versions 3.x and 4.x.
